src/coreclr/jit/codegenxarch.cpp

```diff
@@ -6308,6 +6308,30 @@ void CodeGen::genIntToIntCast(GenTreeCast* cast)
         genIntCastOverflowCheck(cast, desc, srcReg);
     }
 
+    // Code size optimization: use "cdqe" and "cdwe" instead
+    // of "movsxd rax, eax" and "movsx eax, ax", respectively.
+    if ((srcReg == REG_EAX) && (srcReg == dstReg))
+    {
+        emitAttr attribute = EA_UNKNOWN;
+#ifdef TARGET_64BIT
+        if (desc.ExtendKind() == GenIntCastDesc::SIGN_EXTEND_INT)
+        {
+            attribute = EA_8BYTE;
+        }
+#endif
+        if ((desc.ExtendKind() == GenIntCastDesc::SIGN_EXTEND_SMALL_INT) && (desc.ExtendSrcSize() == sizeof(int16_t)))
+        {
+            attribute = EA_4BYTE;
+        }
+
+        if (attribute != EA_UNKNOWN)
+        {
+            emit->emitIns(INS_cwde, attribute);
+            genProduceReg(cast);
+            return;
+        }
+    }
```

Comment:
This looks correct to me, but I also expect there may be cases that are "missing" due to this only being in `genIntToIntCast`. I wonder if it would be better to treat `cwde` specially in `emit` by special casing the `movsx eax, ax` when doing the size computation and actual byte emission.

Reply:
It seems like general goodness to "simplify" the number of `ins` types we can encounter in `emit` and instead specially handle them when actually doing the output encoding.

That would also avoid the various complications that come from cases like `imul_*` and a few other scenarios that have been known to cause problems in the past (even recently).

Reply:
I agree with Tanner here. IMO, a better place for the code size optimization would be https://github.com/dotnet/runtime/blob/fcaba1806105e9e3e097bee4f8a95ccaef79620e/src/coreclr/jit/emitxarch.cpp#L4545

Reply:
I will look into it.